gtsfm/frontend/correspondence_generator/keypoint_aggregator/keypoint_aggregator_dedup.py

```diff
@@ -0,0 +1,106 @@
+"""Keypoint aggregator that de-duplicates keypoint detections, aggregating from image pairs to single images.
+
+Authors: John Lambert
+"""
+
+from typing import Dict, List, Optional, Tuple
+
+import numpy as np
+
+import gtsfm.utils.logger as logger_utils
+from gtsfm.common.keypoints import Keypoints
+from gtsfm.frontend.correspondence_generator.keypoint_aggregator.keypoint_aggregator_base import KeypointAggregatorBase
+
+logger = logger_utils.get_logger()
+
+
+class KeypointAggregatorDedup(KeypointAggregatorBase):
+    """Keypoint aggregator with de-duplication."""
+
+    def run(self, keypoints_dict: Dict[Tuple[int, int], Tuple[Keypoints, Keypoints]]) -> List[Optional[Keypoints]]:
+        """Aggregates per-pair image keypoints into a set of keypoints per image, with de-duplication.
+
+        Keypoints are computed per image pair, instead of per image, so they are aggregated per image here.
+
+        Args:
+            keypoints_dict: key (i1,i2) maps to (keypoints_i1, keypoints_i2) representing matches (correspondences).
+
+        Returns:
+            keypoints_list: list of N Keypoints objects for N images.
+            putative_corr_idxs_dict: mapping from image pair (i1,i2) to putative correspondence indices.
+              Correspondence indices are represented by an array of shape (K,2), for K correspondences.
+        """
+        image_indices = set()
```

Comment:
do you want to break this up into smaller functions? seems like too much code in one method.

Reply:
Agreed. Broke it up now into modular blocks, and removed the code duplication.